daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala

```diff
@@ -149,29 +147,29 @@ object TransactionCoder {
     }
     node match {
       case nc @ NodeCreate(_, _, _, _, _, _, _) =>
-        encodeContractInstance(encodeCid, nc.coinst).flatMap { inst =>
-          val createBuilder = TransactionOuterClass.NodeCreate
+        val createBuilder =
+          TransactionOuterClass.NodeCreate
             .newBuilder()
-            .setContractIdOrStruct(encodeCid, transactionVersion, nc.coid)(
-              _.setContractId(_),
-              _.setContractIdStruct(_),
-            )
-            .setContractInstance(inst)
             .addAllStakeholders(nc.stakeholders.toSet[String].asJava)
             .addAllSignatories(nc.signatories.toSet[String].asJava)
-          nc.key match {
-            case None => Right(nodeBuilder.setCreate(createBuilder).build())
+
+        for {
+          encodedCid <- encodeCid.encode(transactionVersion, nc.coid)
+          _ = encodedCid.fold(createBuilder.setContractId, createBuilder.setContractIdStruct)
+          inst <- encodeContractInstance(encodeCid, nc.coinst)
+          _ = createBuilder.setContractInstance(inst)
+          optKey <- nc.key match {
+            case None => Right(None)
             case Some(key) =>
               if (transactionVersion precedes minKeyOrLookupByKey)
                 Left(EncodeError(transactionVersion, isTooOldFor = "NodeCreate's `key` field"))
               else
-                encodeKeyWithMaintainers(encodeCid, key).map {
-                  case (_, encodedKey) =>
-                    createBuilder.setKeyWithMaintainers(encodedKey)
-                    nodeBuilder.setCreate(createBuilder).build()
-                }
+                encodeKeyWithMaintainers(encodeCid, key).map(Some(_))
+          }
+          _ = optKey.foreach {
```

Comment:
All the `_ = ... createBuilder.setXyz` can be moved to the yield block.

Reply:
OK. Is it really better ?

Reply:
For starters, you don't do work in case a generator downstream goes to `Left`.

Reply:
fair enough.